[`Pyflakes`] `F504` fix changes behavior if args have side effects

### Summary

The [percent-format-extra-named-arguments (F504)](https://docs.astral.sh/ruff/rules/percent-format-extra-named-arguments/#percent-format-extra-named-arguments-f504) fix can change behavior if the extra args have side effects. It should probably be an always unsafe fix, or only safe if the args being removed are only literals.
[playground](https://play.ruff.rs/e87d8cdc-5f74-4e38-bde5-a8951274d82e)
```
PS ~\Desktop\New_folder\ruff>Get-Content issue.py
```
```py
"Hello, %(name)s" % {"greeting": print(1), "name": "World"}
```
```
PS ~\Desktop\New_folder\ruff>py issue.py
```
```
1
```
```
PS ~\Desktop\New_folder\ruff>uvx ruff check issue.py --isolated --select F
```
```snap
issue.py:1:1: F504 [*] `%`-format string has unused named argument(s): greeting
  |
1 | "Hello, %(name)s" % {"greeting": print(1), "name": "World"}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ F504
  |
  = help: Remove extra named arguments: greeting

Found 1 error.
[*] 1 fixable with the `--fix` option.
```
```
PS ~\Desktop\New_folder\ruff>uvx ruff check issue.py --isolated --select F --fix
```
```
Found 1 error (1 fixed, 0 remaining).
```
```
PS ~\Desktop\New_folder\ruff>Get-Content issue.py
```
```py
"Hello, %(name)s" % {"name": "World"}
```
```
PS ~\Desktop\New_folder\ruff>py issue.py
```
```
```

This also applies to [string-dot-format-extra-named-arguments (F522)](https://docs.astral.sh/ruff/rules/string-dot-format-extra-named-arguments/#string-dot-format-extra-named-arguments-f522) [playground](https://play.ruff.rs/55f7e08b-c4f7-4cfc-9512-11109d808095)

This also applies to [string-dot-format-extra-positional-arguments (F523)](https://docs.astral.sh/ruff/rules/string-dot-format-extra-positional-arguments/#string-dot-format-extra-positional-arguments-f523) [playground](https://play.ruff.rs/d586f954-6a06-4ebb-b1b5-d8199b523a68)

### Version

ruff 0.12.0 (87f0feb21 2025-06-17) + playground

--- a/crates/ruff_linter/resources/test/fixtures/pyflakes/F504.py
+++ b/crates/ruff_linter/resources/test/fixtures/pyflakes/F504.py
@@ -14,3 +14,6 @@ hidden = {"a": "!"}
 "" % {
   'test1': '',  'test2': '',
 }
+
+# https://github.com/astral-sh/ruff/issues/18806
+"Hello, %(name)s" % {"greeting": print(1), "name": "World"}

--- a/crates/ruff_linter/resources/test/fixtures/pyflakes/F522.py
+++ b/crates/ruff_linter/resources/test/fixtures/pyflakes/F522.py
@@ -4,3 +4,6 @@
 "{bar:{spam}}".format(bar=2, spam=3, eggs=4, ham=5)  # F522
 (''
  .format(x=2))  # F522
+
+# https://github.com/astral-sh/ruff/issues/18806
+"Hello, {name}".format(greeting=print(1), name="World")

--- a/crates/ruff_linter/resources/test/fixtures/pyflakes/F522.py
+++ b/crates/ruff_linter/resources/test/fixtures/pyflakes/F522.py
@@ -6,4 +6,9 @@
  .format(x=2))  # F522
 
 # https://github.com/astral-sh/ruff/issues/18806
+# The fix here is unsafe because the unused argument has side effect
 "Hello, {name}".format(greeting=print(1), name="World")
+
+# The fix here is safe because the unused argument has no side effect,
+# even though the used argument has a side effect
+"Hello, {name}".format(greeting="Pikachu", name=print(1))

--- a/crates/ruff_linter/resources/test/fixtures/pyflakes/F523.py
+++ b/crates/ruff_linter/resources/test/fixtures/pyflakes/F523.py
@@ -35,3 +35,6 @@
 # Removing the final argument.
 "Hello".format("world")
 "Hello".format("world", key="value")
+
+# https://github.com/astral-sh/ruff/issues/18806
+"Hello, {0}".format("world", print(1))

--- a/crates/ruff_linter/resources/test/fixtures/pyflakes/F523.py
+++ b/crates/ruff_linter/resources/test/fixtures/pyflakes/F523.py
@@ -37,4 +37,9 @@
 "Hello".format("world", key="value")
 
 # https://github.com/astral-sh/ruff/issues/18806
+# The fix here is unsafe because the unused argument has side effect
 "Hello, {0}".format("world", print(1))
+
+# The fix here is safe because the unused argument has no side effect,
+# even though the used argument has a side effect
+"Hello, {0}".format(print(1), "Pikachu")

--- a/crates/ruff_linter/src/rules/pyflakes/rules/strings.rs
+++ b/crates/ruff_linter/src/rules/pyflakes/rules/strings.rs
@@ -777,14 +777,17 @@ pub(crate) fn string_dot_format_extra_named_arguments(
 
     let keyword_names = keywords
         .iter()
-        .filter_map(|Keyword { arg, .. }| arg.as_ref());
+        .filter_map(|Keyword { arg, value, .. }| Some((arg.as_ref()?, value)));
 
+    let mut side_effects = false;
     let missing: Vec<(usize, &Name)> = keyword_names
         .enumerate()
-        .filter_map(|(index, keyword)| {
+        .filter_map(|(index, (keyword, value))| {
             if summary.keywords.contains(keyword.id()) {
                 None
             } else {
+                side_effects |=
+                    contains_effect(value, |id| checker.semantic().has_builtin_binding(id));
                 Some((index, &keyword.id))
             }
         })
@@ -799,7 +802,7 @@ pub(crate) fn string_dot_format_extra_named_arguments(
         StringDotFormatExtraNamedArguments { missing: names },
         call.range(),
     );
-    let indexes: Vec<usize> = missing.iter().map(|(index, _)| *index).collect();
+    let indexes: Vec<usize> = missing.into_iter().map(|(index, _)| index).collect();
     diagnostic.try_set_fix(|| {
         let edit = remove_unused_keyword_arguments_from_format_call(
             &indexes,
@@ -811,13 +814,7 @@ pub(crate) fn string_dot_format_extra_named_arguments(
         Ok(Fix::applicable_edit(
             edit,
             // Mark fix as unsafe if the `format` call contains an argument with side effect
-            if keywords
-                .iter()
-                .enumerate()
-                .filter(|(i, _)| indexes.contains(i))
-                .map(|(_, keyword)| &keyword.value)
-                .any(|expr| contains_effect(expr, |id| checker.semantic().has_builtin_binding(id)))
-            {
+            if side_effects {
                 Applicability::Unsafe
             } else {
                 Applicability::Safe
@@ -858,13 +855,17 @@ pub(crate) fn string_dot_format_extra_positional_arguments(
         true
     }
 
+    let mut side_effects = false;
     let missing: Vec<usize> = args
         .iter()
         .enumerate()
         .filter(|(i, arg)| {
             !(arg.is_starred_expr() || summary.autos.contains(i) || summary.indices.contains(i))
         })
-        .map(|(i, _)| i)
+        .map(|(i, arg)| {
+            side_effects |= contains_effect(arg, |id| checker.semantic().has_builtin_binding(id));
+            i
+        })
         .collect();
 
     if missing.is_empty() {
@@ -892,14 +893,7 @@ pub(crate) fn string_dot_format_extra_positional_arguments(
             Ok(Fix::applicable_edit(
                 edit,
                 // Mark fix as unsafe if the `format` call contains an argument with side effect
-                if args
-                    .iter()
-                    .enumerate()
-                    .filter(|(i, _)| missing.contains(i))
-                    .any(|(_, expr)| {
-                        contains_effect(expr, |id| checker.semantic().has_builtin_binding(id))
-                    })
-                {
+                if side_effects {
                     Applicability::Unsafe
                 } else {
                     Applicability::Safe

--- a/crates/ruff_linter/src/rules/pyflakes/rules/strings.rs
+++ b/crates/ruff_linter/src/rules/pyflakes/rules/strings.rs
@@ -140,6 +140,15 @@ impl Violation for PercentFormatExpectedSequence {
 /// "Hello, %(name)s" % {"name": "World"}
 /// ```
 ///
+/// ## Fix safety
+/// This rule's fix is marked as unsafe if there's a call expression in the
+/// `printf`-style format string.
+///
+/// For example, the fix would be marked as unsafe in the following case:
+/// ```python
+/// "Hello, %(name)s" % {"greeting": print(1), "name": "World"}
+/// ```
+///
 /// ## References
 /// - [Python documentation: `printf`-style String Formatting](https://docs.python.org/3/library/stdtypes.html#printf-style-string-formatting)
 #[derive(ViolationMetadata)]
@@ -381,6 +390,15 @@ impl Violation for StringDotFormatInvalidFormat {
 /// "Hello, {name}".format(name="World")
 /// ```
 ///
+/// ## Fix safety
+/// This rule's fix is marked as unsafe if there's a call expression in the
+/// `format` call.
+///
+/// For example, the fix would be marked as unsafe in the following case:
+/// ```python
+/// "Hello, {name}".format(greeting=print(1), name="World")
+/// ```
+///
 /// ## References
 /// - [Python documentation: `str.format`](https://docs.python.org/3/library/stdtypes.html#str.format)
 #[derive(ViolationMetadata)]
@@ -420,6 +438,15 @@ impl Violation for StringDotFormatExtraNamedArguments {
 /// Use instead:
 /// ```python
 /// "Hello, {0}".format("world")
+/// ```
+///
+/// ## Fix safety
+/// This rule's fix is marked as unsafe if there's a call expression in the
+/// `format` call.
+///
+/// For example, the fix would be marked as unsafe in the following case:
+/// ```python
+/// "Hello, {0}".format("world", print(1))
 /// ```
 ///
 /// ## References

--- a/crates/ruff_linter/src/rules/pyflakes/rules/strings.rs
+++ b/crates/ruff_linter/src/rules/pyflakes/rules/strings.rs
@@ -141,8 +141,9 @@ impl Violation for PercentFormatExpectedSequence {
 /// ```
 ///
 /// ## Fix safety
-/// This rule's fix is marked as unsafe if there's a call expression in the
-/// `printf`-style format string.
+/// This rule's fix is marked as unsafe for mapping key
+/// containing function calls with potential side effects,
+/// because removing such arguments could change the behavior of the code.
 ///
 /// For example, the fix would be marked as unsafe in the following case:
 /// ```python
@@ -391,8 +392,9 @@ impl Violation for StringDotFormatInvalidFormat {
 /// ```
 ///
 /// ## Fix safety
-/// This rule's fix is marked as unsafe if there's a call expression in the
-/// `format` call.
+/// This rule's fix is marked as unsafe if the unused keyword argument
+/// contains a function call with potential side effects,
+/// because removing such arguments could change the behavior of the code.
 ///
 /// For example, the fix would be marked as unsafe in the following case:
 /// ```python
@@ -441,8 +443,9 @@ impl Violation for StringDotFormatExtraNamedArguments {
 /// ```
 ///
 /// ## Fix safety
-/// This rule's fix is marked as unsafe if there's a call expression in the
-/// `format` call.
+/// This rule's fix is marked as unsafe if the unused positional argument
+/// contains a function call with potential side effects,
+/// because removing such arguments could change the behavior of the code.
 ///
 /// For example, the fix would be marked as unsafe in the following case:
 /// ```python

--- a/crates/ruff_linter/src/rules/pyflakes/rules/strings.rs
+++ b/crates/ruff_linter/src/rules/pyflakes/rules/strings.rs
@@ -1,7 +1,7 @@
 use std::string::ToString;
 
 use ruff_diagnostics::Applicability;
-use ruff_python_parser::TokenKind;
+use ruff_python_ast::helpers::contains_effect;
 use rustc_hash::FxHashSet;
 
 use ruff_macros::{ViolationMetadata, derive_message_formats};
@@ -643,8 +643,8 @@ pub(crate) fn percent_format_extra_named_arguments(
         )?;
         Ok(Fix::applicable_edit(
             edit,
-            // Mark fix as unsafe if `dict` contains a call expression
-            if contains_call_expr_in_range(checker, dict.range()) {
+            // Mark fix as unsafe if `dict` contains a call with side effect
+            if contains_effect(right, |id| checker.semantic().has_builtin_binding(id)) {
                 Applicability::Unsafe
             } else {
                 Applicability::Safe
@@ -806,8 +806,10 @@ pub(crate) fn string_dot_format_extra_named_arguments(
         )?;
         Ok(Fix::applicable_edit(
             edit,
-            // Mark fix as unsafe if the call arguments contains a call expression
-            if contains_call_expr_in_range(checker, call.arguments.range().add_start(1.into())) {
+            // Mark fix as unsafe if the call arguments contains a call with side effect
+            if call.arguments.arguments_source_order().any(|arg| {
+                contains_effect(arg.value(), |id| checker.semantic().has_builtin_binding(id))
+            }) {
                 Applicability::Unsafe
             } else {
                 Applicability::Safe
@@ -881,9 +883,10 @@ pub(crate) fn string_dot_format_extra_positional_arguments(
             )?;
             Ok(Fix::applicable_edit(
                 edit,
-                // Mark fix as unsafe if the call arguments contains a call expression
-                if contains_call_expr_in_range(checker, call.arguments.range().add_start(1.into()))
-                {
+                // Mark fix as unsafe if the call arguments contains a call with side effect
+                if call.arguments.arguments_source_order().any(|arg| {
+                    contains_effect(arg.value(), |id| checker.semantic().has_builtin_binding(id))
+                }) {
                     Applicability::Unsafe
                 } else {
                     Applicability::Safe
@@ -942,13 +945,4 @@ pub(crate) fn string_dot_format_mixing_automatic(
     if !(summary.autos.is_empty() || summary.indices.is_empty()) {
         checker.report_diagnostic(StringDotFormatMixingAutomatic, call.range());
     }
-}
-
-/// Searches for a call expression by looking for a [`TokenKind::Lpar`] token.
-fn contains_call_expr_in_range(checker: &Checker, range: TextRange) -> bool {
-    checker
-        .tokens()
-        .in_range(range)
-        .iter()
-        .any(|token| token.kind() == TokenKind::Lpar)
 }

--- a/crates/ruff_linter/src/rules/pyflakes/rules/strings.rs
+++ b/crates/ruff_linter/src/rules/pyflakes/rules/strings.rs
@@ -1,5 +1,7 @@
 use std::string::ToString;
 
+use ruff_diagnostics::Applicability;
+use ruff_python_parser::TokenKind;
 use rustc_hash::FxHashSet;
 
 use ruff_macros::{ViolationMetadata, derive_message_formats};
@@ -603,15 +605,24 @@ pub(crate) fn percent_format_extra_named_arguments(
         PercentFormatExtraNamedArguments { missing: names },
         location,
     );
-    let indexes: Vec<usize> = missing.iter().map(|(index, _)| *index).collect();
+
     diagnostic.try_set_fix(|| {
+        let indexes: Vec<usize> = missing.iter().map(|(index, _)| *index).collect();
         let edit = remove_unused_format_arguments_from_dict(
             &indexes,
             dict,
             checker.locator(),
             checker.stylist(),
         )?;
-        Ok(Fix::safe_edit(edit))
+        Ok(Fix::applicable_edit(
+            edit,
+            // Mark fix as unsafe if `dict` contains a call expression
+            if contains_call_expr_in_range(checker, dict.range()) {
+                Applicability::Unsafe
+            } else {
+                Applicability::Safe
+            },
+        ))
     });
 }
 
@@ -766,7 +777,15 @@ pub(crate) fn string_dot_format_extra_named_arguments(
             checker.locator(),
             checker.stylist(),
         )?;
-        Ok(Fix::safe_edit(edit))
+        Ok(Fix::applicable_edit(
+            edit,
+            // Mark fix as unsafe if the call arguments contains a call expression
+            if contains_call_expr_in_range(checker, call.arguments.range().add_start(1.into())) {
+                Applicability::Unsafe
+            } else {
+                Applicability::Safe
+            },
+        ))
     });
 }
 
@@ -833,7 +852,16 @@ pub(crate) fn string_dot_format_extra_positional_arguments(
                 checker.locator(),
                 checker.stylist(),
             )?;
-            Ok(Fix::safe_edit(edit))
+            Ok(Fix::applicable_edit(
+                edit,
+                // Mark fix as unsafe if the call arguments contains a call expression
+                if contains_call_expr_in_range(checker, call.arguments.range().add_start(1.into()))
+                {
+                    Applicability::Unsafe
+                } else {
+                    Applicability::Safe
+                },
+            ))
         });
     }
 }
@@ -887,4 +915,13 @@ pub(crate) fn string_dot_format_mixing_automatic(
     if !(summary.autos.is_empty() || summary.indices.is_empty()) {
         checker.report_diagnostic(StringDotFormatMixingAutomatic, call.range());
     }
+}
+
+/// Searches for a call expression by looking for a [`TokenKind::Lpar`] token.
+fn contains_call_expr_in_range(checker: &Checker, range: TextRange) -> bool {
+    checker
+        .tokens()
+        .in_range(range)
+        .iter()
+        .any(|token| token.kind() == TokenKind::Lpar)
 }

--- a/crates/ruff_linter/src/rules/pyflakes/rules/strings.rs
+++ b/crates/ruff_linter/src/rules/pyflakes/rules/strings.rs
@@ -775,11 +775,11 @@ pub(crate) fn string_dot_format_extra_named_arguments(
         return;
     }
 
-    let keywords = keywords
+    let keyword_names = keywords
         .iter()
         .filter_map(|Keyword { arg, .. }| arg.as_ref());
 
-    let missing: Vec<(usize, &Name)> = keywords
+    let missing: Vec<(usize, &Name)> = keyword_names
         .enumerate()
         .filter_map(|(index, keyword)| {
             if summary.keywords.contains(keyword.id()) {
@@ -807,12 +807,17 @@ pub(crate) fn string_dot_format_extra_named_arguments(
             checker.locator(),
             checker.stylist(),
         )?;
+
         Ok(Fix::applicable_edit(
             edit,
-            // Mark fix as unsafe if the call arguments contains a call with side effect
-            if call.arguments.arguments_source_order().any(|arg| {
-                contains_effect(arg.value(), |id| checker.semantic().has_builtin_binding(id))
-            }) {
+            // Mark fix as unsafe if the `format` call contains an argument with side effect
+            if keywords
+                .iter()
+                .enumerate()
+                .filter(|(i, _)| indexes.contains(i))
+                .map(|(_, keyword)| &keyword.value)
+                .any(|expr| contains_effect(expr, |id| checker.semantic().has_builtin_binding(id)))
+            {
                 Applicability::Unsafe
             } else {
                 Applicability::Safe
@@ -886,10 +891,15 @@ pub(crate) fn string_dot_format_extra_positional_arguments(
             )?;
             Ok(Fix::applicable_edit(
                 edit,
-                // Mark fix as unsafe if the call arguments contains a call with side effect
-                if call.arguments.arguments_source_order().any(|arg| {
-                    contains_effect(arg.value(), |id| checker.semantic().has_builtin_binding(id))
-                }) {
+                // Mark fix as unsafe if the `format` call contains an argument with side effect
+                if args
+                    .iter()
+                    .enumerate()
+                    .filter(|(i, _)| missing.contains(i))
+                    .any(|(_, expr)| {
+                        contains_effect(expr, |id| checker.semantic().has_builtin_binding(id))
+                    })
+                {
                     Applicability::Unsafe
                 } else {
                     Applicability::Safe

